File: src/main/java/org/lanternpowered/server/service/permission/LanternPermissionDescription.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.service.permission;

import static com.google.common.base.Preconditions.checkNotNull;
import static com.google.common.base.Preconditions.checkState;

import com.google.common.base.MoreObjects;
import com.google.common.base.Objects;
import com.google.common.base.Preconditions;
import org.lanternpowered.server.service.permission.base.LanternSubjectCollection;
import org.spongepowered.api.plugin.PluginContainer;
import org.spongepowered.api.service.permission.PermissionDescription;
import org.spongepowered.api.service.permission.PermissionService;
import org.spongepowered.api.service.permission.Subject;
import org.spongepowered.api.service.permission.SubjectCollection;
import org.spongepowered.api.service.permission.SubjectData;
import org.spongepowered.api.service.permission.SubjectReference;
import org.spongepowered.api.text.Text;
import org.spongepowered.api.util.Tristate;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

import org.checkerframework.checker.nullness.qual.Nullable;

/**
 * Basic implementation of {@link PermissionDescription}. Can only be used in
 * conjunction with {@link LanternPermissionService}.
 */
final class LanternPermissionDescription implements PermissionDescription {

    private final LanternPermissionService permissionService;
    private final PluginContainer owner;
    private final String id;
    @Nullable private final Text description;

    private LanternPermissionDescription(LanternPermissionService permissionService, String id, PluginContainer owner, @Nullable Text description) {
        super();
        this.permissionService = permissionService;
        this.description = description;
        this.owner = owner;
        this.id = id;
    }

    @Override
    public String getId() {
        return this.id;
    }

    @Override
    public Optional<Text> getDescription() {
        return Optional.ofNullable(this.description);
    }

    @Override
    public Map<Subject, Boolean> getAssignedSubjects(String identifier) {
        final SubjectCollection subjects = this.permissionService.get(identifier);
        return subjects.getLoadedWithPermission(this.id);
    }

    @Override
    public CompletableFuture<Map<SubjectReference, Boolean>> findAssignedSubjects(String type) {
        final SubjectCollection subjects = this.permissionService.get(type);
        return subjects.getAllWithPermission(this.id);
    }

    @Override
    public Optional<PluginContainer> getOwner() {
        return Optional.of(this.owner);
    }

    @Override
    public int hashCode() {
        return this.id.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final LanternPermissionDescription other = (LanternPermissionDescription) obj;
        return this.id.equals(other.id) && this.owner.equals(other.owner) && Objects.equal(this.description, other.description);
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("owner", this.owner)
                .add("id", this.id)
                .toString();
    }

    static class Builder implements PermissionDescription.Builder {

        private final Map<String, Tristate> roleAssignments = new LinkedHashMap<>();
        private final LanternPermissionService permissionService;
        private final PluginContainer owner;
        @Nullable private Text description;
        @Nullable private String id;

        Builder(LanternPermissionService permissionService, PluginContainer owner) {
            super();
            this.permissionService = permissionService;
            this.owner = owner;
        }

        @Override
        public Builder id(String id) {
            this.id = checkNotNull(id, "permissionId");
            return this;
        }

        @Override
        public Builder description(@Nullable Text description) {
            this.description = description;
            return this;
        }

        @Override
        public Builder assign(String role, boolean value) {
            Preconditions.checkNotNull(role, "role");
            this.roleAssignments.put(role, Tristate.fromBoolean(value));
            return this;
        }

        @Override
        public LanternPermissionDescription register() throws IllegalStateException {
            checkState(this.id != null, "No id set");
            final LanternPermissionDescription description = new LanternPermissionDescription(
                    this.permissionService, this.id, this.owner, this.description);
            this.permissionService.addDescription(description);

            // Set role-templates
            final LanternSubjectCollection subjects = this.permissionService.get(PermissionService.SUBJECTS_ROLE_TEMPLATE);
            for (Entry<String, Tristate> assignment : this.roleAssignments.entrySet()) {
                final Subject subject = subjects.get(assignment.getKey());
                subject.getTransientSubjectData().setPermission(SubjectData.GLOBAL_CONTEXT, this.id, assignment.getValue());
            }
            return description;
        }
    }

}

File: src/main/kotlin/org/lanternpowered/server/service/permission/LanternPermissionDescription.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.service.permission

import org.lanternpowered.api.text.Text
import org.lanternpowered.api.util.ToStringHelper
import org.lanternpowered.api.util.optional.asOptional
import org.spongepowered.api.service.permission.PermissionDescription
import org.spongepowered.api.service.permission.PermissionService
import org.spongepowered.api.service.permission.Subject
import org.spongepowered.api.service.permission.SubjectData
import org.spongepowered.api.service.permission.SubjectReference
import org.spongepowered.api.util.Tristate
import org.spongepowered.plugin.PluginContainer
import java.util.Optional
import java.util.concurrent.CompletableFuture

/**
 * Basic implementation of [PermissionDescription]. Can only be used in
 * conjunction with [LanternPermissionService].
 */
internal class LanternPermissionDescription private constructor(
        private val permissionService: LanternPermissionService,
        private val id: String,
        private val owner: PluginContainer,
        private val description: Text?
) : PermissionDescription {

    override fun getId(): String = this.id
    override fun getDescription(): Optional<Text> = this.description.asOptional()

    override fun getAssignedSubjects(identifier: String): Map<Subject, Boolean> {
        val subjects = this.permissionService[identifier]
        return subjects.getLoadedWithPermission(this.id)
    }

    override fun findAssignedSubjects(type: String): CompletableFuture<Map<SubjectReference, Boolean>> {
        val subjects = this.permissionService[type]
        return subjects.getAllWithPermission(this.id)
    }

    override fun getOwner(): Optional<PluginContainer> = this.owner.asOptional()
    override fun hashCode(): Int = this.id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other)
            return true
        if (other == null)
            return false
        if (this.javaClass != other.javaClass) {
            return false
        }
        other as LanternPermissionDescription
        return id == other.id && owner == other.owner && description == other.description
    }

    override fun toString(): String = ToStringHelper(this)
            .add("owner", this.owner)
            .add("id", this.id)
            .toString()

    internal class Builder(
            private val permissionService: LanternPermissionService,
            private val owner: PluginContainer
    ) : PermissionDescription.Builder {

        private val roleAssignments = linkedMapOf<String, Tristate>()
        private var description: Text? = null
        private var id: String? = null

        override fun id(id: String): Builder = apply { this.id = id }
        override fun description(description: Text?): Builder = apply { this.description = description }

        override fun assign(role: String, value: Boolean): Builder {
            this.roleAssignments[role] = Tristate.fromBoolean(value)
            return this
        }

        override fun register(): LanternPermissionDescription {
            val id = checkNotNull(this.id) { "No id set" }
            val description = LanternPermissionDescription(this.permissionService, id, this.owner, this.description)
            this.permissionService.addDescription(description)

            // Set role-templates
            val subjects = this.permissionService[PermissionService.SUBJECTS_ROLE_TEMPLATE]
            for ((key, value) in this.roleAssignments)
                subjects[key].transientSubjectData.setPermission(SubjectData.GLOBAL_CONTEXT, id, value)
            return description
        }
    }
}
